ompi/mca/coll/acoll/coll_acoll_allgather.c

```diff
@@ -0,0 +1,625 @@
+                                MPI_STATUS_IGNORE));
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+    }
+
+    return err;
+}
+
+/*
+ * sg_bcast_intra
+ *
+ * Function:    broadcast operation within a subgroup
+ * Accepts:     Arguments of MPI_Bcast() plus subgroup params
+ * Returns:     MPI_SUCCESS or error code
+ *
+ * Description: O(N) or O(log(N)) algorithm based on count.
+ *
+ * Memory:      No additional memory requirements beyond user-supplied buffers.
+ *
+ */
+static inline int sg_bcast_intra(void *buff, size_t count, struct ompi_datatype_t *datatype, int rank,
+                                 int dim, int size, int sg_size, int cur_base, int sg_start,
+                                 struct ompi_communicator_t *comm, mca_coll_base_module_t *module,
+                                 ompi_request_t **preq, int *nreqs)
+{
+    int err;
+    size_t total_dsize, dsize;
+
+    ompi_datatype_type_size(datatype, &dsize);
+    total_dsize = dsize * count;
+
+    if (total_dsize <= 8192) {
+        err = log_sg_bcast_intra(buff, count, datatype, rank, dim, size, sg_size, cur_base,
+                                 sg_start, comm, module, preq, nreqs);
+    } else {
+        err = lin_sg_bcast_intra(buff, count, datatype, rank, dim, size, sg_size, cur_base,
+                                 sg_start, comm, module, preq, nreqs);
+    }
+    return err;
+}
+
+/*
+ * coll_allgather_decision_fixed
+ *
+ * Function:    Choose optimal allgather algorithm
+ *
+ * Description: Based on no. of processes and message size, chooses whether
+ *              or not to use subgroups. If subgroup based algorithm is not,
+ *              chosen, further decides if [ring|lin] allgather is to be used.
+ *
+ */
+static inline void coll_allgather_decision_fixed(int size, size_t total_dsize, int sg_size,
+                                                 int *use_ring, int *use_lin)
+{
+    *use_ring = 0;
+    *use_lin = 0;
+    if (size <= (sg_size << 1)) {
+        if (total_dsize >= 1048576) {
+            *use_lin = 1;
+        }
+    } else if (size <= (sg_size << 2)) {
+        if ((total_dsize >= 4096) && (total_dsize < 32768)) {
+            *use_ring = 1;
+        } else if (total_dsize >= 1048576) {
+            *use_lin = 1;
+        }
+    } else if (size <= (sg_size << 3)) {
+        if ((total_dsize >= 4096) && (total_dsize < 32768)) {
+            *use_ring = 1;
+        }
+    } else {
+        if (total_dsize >= 4096) {
+            *use_ring = 1;
+        }
+    }
+}
+
+/*
+ * rd_allgather_sub
+ *
+ * Function:    Uses recursive doubling based allgather for the group.
+ *              Group can be all ranks in a subgroup or base ranks across
+ *              subgroups.
+ *
+ * Description: Implementation logic of recursive doubling reused from
+ *              ompi_coll_base_allgather_intra_recursivedoubling().
+ *
+ */
+static inline int rd_allgather_sub(void *rbuf, struct ompi_datatype_t *rdtype,
```

Comment:
Why doesn't this function take sbuf?

Reply:
The relevant data from `sbuf` is copied to `rbuf` at the beginning of `mca_coll_acoll_allgather_intra()`, so it does not need `sbuf`.